compiler/damlc/daml-doc/src/DA/Daml/Doc/Render/Markdown.hs

```diff
@@ -9,98 +9,124 @@ module DA.Daml.Doc.Render.Markdown
-  , if null md_classes
-    then []
-    else [ "## Typeclasses"
-         , ""
-         , T.unlines $ map cls2md md_classes
-         , ""
-         ]
-  , if null md_adts then []
-    else [ "## Data types"
-         , ""
-         , T.unlines $ map adt2md md_adts
-         , "" ]
-  , if null md_functions then []
-    else [ "## Functions"
-         , ""
-         , T.unlines $ map fct2md md_functions
-         ]
-  ]
-
-
-tmpl2md :: TemplateDoc -> T.Text
-tmpl2md TemplateDoc{..} = T.unlines $
-    [ "### Template " <> asCode (unTypename td_name)
-    , maybe "" (T.cons '\n' . unDocText) td_descr
-    , ""
-    , fieldTable td_payload
-    , ""
-    , "  #### Choices"
-    , ""
-    ] ++ map choiceBullet td_choices -- ends by "\n" because of unlines above
-
+    isNothing md_descr = mempty
+renderSimpleMD ModuleDoc{..} = mconcat
+    [ renderAnchorInfix "# " md_anchor ("Module " <> escapeMd (unModulename md_name))
+    , renderDocText md_descr
+    , section "Templates" tmpl2md md_templates
+    , section "Typeclasses" cls2md md_classes
+    , section "Data types" adt2md md_adts
+    , section "Functions" fct2md md_functions
+    ]
   where
-    choiceBullet :: ChoiceDoc -> T.Text
-    choiceBullet ChoiceDoc{..} = T.unlines
-        [ prefix "* " $ asCode (unTypename cd_name)
-        , maybe "  " (flip T.snoc '\n' . indent 2 . unDocText) cd_descr
-        , indent 2 (fieldTable cd_fields)
+    section :: T.Text -> (a -> RenderOut) -> [a] -> RenderOut
+    section _ _ [] = mempty
+    section sectionTitle f xs = mconcat
+        [ renderLines
+            [ "## " <> sectionTitle
+            , ""
+            ]
+        , mconcatMap f xs
+        , renderLine ""
         ]
 
-cls2md :: ClassDoc -> T.Text
-cls2md ClassDoc{..} = T.unlines $
-    [ "### `class` "
-        <> maybe "" (\x -> type2md x <> " => ") cl_super
-        <> T.unwords (unTypename cl_name : cl_args)
-        <> " where"
-    , maybe "" (T.cons '\n' . indent 2 . unDocText) cl_descr
-    ] ++ map (indent 2 . fct2md) cl_functions
-
-adt2md :: ADTDoc -> T.Text
-adt2md TypeSynDoc{..} = T.unlines $
-    [ "### `type` "
-        <> asCode (unTypename ad_name <> (T.concat $ map (T.cons ' ') ad_args))
-    , "    = " <> type2md ad_rhs
-    ] ++ maybe [] ((:[]) . T.cons '\n' . indent 2 . unDocText) ad_descr
-
-adt2md ADTDoc{..} = T.unlines $
-    [ "### `data` "
-        <> asCode (unTypename ad_name <> (T.concat $ map (T.cons ' ') ad_args))
-    , maybe T.empty (T.cons '\n' . indent 2 . unDocText) ad_descr
-    ] ++ map constrMdItem ad_constrs
-
-constrMdItem :: ADTConstr -> T.Text
-constrMdItem PrefixC{..} =
-  ("* " <> T.unwords (asCode (unTypename ac_name) : map type2md ac_args))
-  <> maybe T.empty (T.cons '\n' . indent 2 . unDocText) ac_descr
-constrMdItem RecordC{..} =
-  ("* " <> asCode (unTypename ac_name))
-  <> maybe T.empty (T.cons '\n' . indent 2 . unDocText) ac_descr
-  <> "\n\n"
-  <> indent 2 (fieldTable ac_fields)
 
+tmpl2md :: TemplateDoc -> RenderOut
```

Comment:
I'm not the biggest fan of abbreviations when the actual name is not much longer. It makes it harder for new people to dive into the code base. Nothing to change now but probably worth consider in the future.